File: yaml_parser/shapes.py
from transform import Transform
from material import Material
from pattern import Pattern

from copy import deepcopy


class Shape(object):
    def __init__(self, yaml_obj, defines=None):
        self.yaml_obj = yaml_obj
        self.defines = defines
        self.name = ...

    @classmethod
    def from_yaml(cls, obj, defines) -> 'Shape':
        if 'transform' not in obj:
            obj['transform'] = []

        if obj["add"] == "sphere":
            return Sphere.from_yaml(obj)
        elif obj["add"] == "plane":
            return Plane.from_yaml(obj)
        elif obj["add"] == "cube":
            return Cube.from_yaml(obj)
        elif obj["add"] == "cone":
            return Cone.from_yaml(obj)
        elif obj["add"] == "cylinder":
            return Cylinder.from_yaml(obj)
        elif obj["add"] == "triangle":
            return Triangle.from_yaml(obj)
        elif obj["add"] == "smooth-triangle":
            return SmoothTriangle.from_yaml(obj)
        elif obj['add'] in ['toroid', 'torus']:
            return Toroid.from_yaml(obj);
        elif obj["add"] == "group":
            if "material" in obj:
                for child in obj["children"]:
                    if "material" not in child:
                        child["material"] = deepcopy(obj["material"])
            return Group.from_yaml(obj, defines)
        elif obj["add"] == "csg":
            if "material" in obj:
                if "material" not in obj["left"]:
                    obj["left"]["material"] = deepcopy(obj["material"])
                if "material" not in obj["right"]:
                    obj["right"]["material"] = deepcopy(obj["material"])
            return CSG.from_yaml(obj, defines)
        elif obj["add"] == "obj":
            from obj_parser import OBJParser
            return OBJParser.from_yaml(obj, defines)

        raise ValueError("unsupported shape.")

    def get_file_name(self):
        if 'file' in self.yaml_obj:
            return self.yaml_obj['file']
        return None

    def c_repr(self, name, parent_name, offset, resources):
        raise NotImplementedError

class CSG(Shape):
    def __init__(self, yaml_obj, defines):
        Shape.__init__(self, yaml_obj, defines)

    @classmethod
    def from_yaml(cls, tree, defines) -> 'CSG':
        return cls(tree, defines)

    def c_repr(self, name, parent_name, offset, resources):
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        left = Shape.from_yaml(self.yaml_obj['left'], self.defines)
        right = Shape.from_yaml(self.yaml_obj['right'], self.defines)

        if 'op' in self.yaml_obj:
            op = self.yaml_obj['op']
        else:
            op = self.yaml_obj['operation']

        if op == 'difference':
            op_str = 'CSG_DIFFERENCE'
        elif op == 'intersection':
            op_str = 'CSG_INTERSECT'
        elif op == 'union':
            op_str = 'CSG_UNION'
        else:
            raise ValueError("Unknown CSG operation: {}".format(op))

        buf = """
    /* children for {0} */
    Shape shape_{0}_children = array_of_shapes({1});
""".format(name, 2)

        buf += """    {0}""".format(left.c_repr('{}_left'.format(name), 'shape_{0}_children'.format(name), 0, resources))

        file_name = left.get_file_name()
        if file_name is not None:
            resources[file_name] = 'shape_{}_left'.format(name)

        buf += """    {0}""".format(right.c_repr('{}_right'.format(name), 'shape_{0}_children'.format(name), 1, resources))

        file_name = right.get_file_name()
        if file_name is not None:
            resources[file_name] = 'shape_{}_right'.format(name)


        buf += """
    /* end children for {0} */

    {2}
    Shape shape_{0} = {1} + {4};
    csg(shape_{0}, {3}, shape_{0}_left, shape_{0}_right);
    shape_set_transform(shape_{0}, transform_{0});
""".format(name,
           parent_name,
           transform.c_repr(name),
           op_str,
           offset)

        return buf


class Group(Shape):
    def __init__(self, yaml_obj, defines):
        Shape.__init__(self, yaml_obj, defines)

    @classmethod
    def from_yaml(cls, tree, defines) -> 'Group':
        return cls(tree, defines)

    def c_repr(self, name, parent_name, offset, resources):
        if 'transform' not in self.yaml_obj:
            yaml_obj['transform'] = []

        transform = Transform.from_yaml(self.yaml_obj['transform'])
        children = []
        for child in self.yaml_obj['children']:
            children.append(Shape.from_yaml(child, self.defines))
            
        buf = """
    /* children for {0} */
    Shape shape_{0}_children = array_of_shapes({1});
""".format(name, len(children))

        for i, child in enumerate(children):
            buf += """
    {0}
""".format(child.c_repr('{0}_child_{1}'.format(name, i), 'shape_{0}_children'.format(name), i, resources))

            file_name = child.get_file_name()
            if file_name is not None:
                resources[file_name] = 'shape_{0}_child_{1}'.format(name, i)
        buf += """    /* end children for {0} */

    {2}
    Shape shape_{0} = {1} + {4};
    group(shape_{0}, shape_{0}_children, {3});
    //shape_free(shape_{0}_children);
    shape_set_transform(shape_{0}, transform_{0});
""".format(name,
          parent_name,
          transform.c_repr(name),
          len(children),
          offset)

        return buf


class Sphere(Shape):
    def __init__(self, yaml_obj):
        Shape.__init__(self, yaml_obj)

    @classmethod
    def from_yaml(cls, obj) -> 'Sphere':
        return cls(obj)

    def c_repr(self, name, parent_name, offset, resources):
        material = Material.from_yaml(self.yaml_obj['material'])
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        buf = """
    {2}
    {3}
    Shape shape_{0} = {1} + {4};
    sphere(shape_{0});
    shape_set_material(shape_{0}, material_{0});
    shape_set_transform(shape_{0}, transform_{0});
""".format(name,
           parent_name,
           material.c_repr(name),
           transform.c_repr(name),
           offset)

        return buf

class Toroid(Shape):
    def __init__(self, yaml_obj):
        Shape.__init__(self, yaml_obj)

    @classmethod
    def from_yaml(cls, obj) -> 'Toroid':
        if 'r1' not in obj:
            obj['r1'] = 0.75
        if 'r2' not in obj:
            obj['r2'] = 0.25
        return cls(obj)

    def c_repr(self, name, parent_name, offset, resources):
        material = Material.from_yaml(self.yaml_obj['material'])
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        buf = """
    {2}
    {3}
    Shape shape_{0} = {1} + {4};
    toroid(shape_{0});
    shape_{0}->fields.toroid.r1 = {5};
    shape_{0}->fields.toroid.r2 = {6};
    shape_set_material(shape_{0}, material_{0});
    shape_set_transform(shape_{0}, transform_{0});
""".format(name,
           parent_name,
           material.c_repr(name),
           transform.c_repr(name),
           offset,
           self.yaml_obj['r1'],
           self.yaml_obj['r2'])

        return buf

class Plane(Shape):
    def __init__(self, yaml_obj):
        Shape.__init__(self, yaml_obj)

    @classmethod
    def from_yaml(cls, obj) -> 'Plane':
        return cls(yaml_obj=obj)

    def c_repr(self, name, parent_name, offset, resources):
        material = Material.from_yaml(self.yaml_obj['material'])
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        buf = """
    {2}
    {3}
    Shape shape_{0} = {1} + {4};
    plane(shape_{0});
    shape_set_material(shape_{0}, material_{0});
    shape_set_transform(shape_{0}, transform_{0});
""".format(name,
           parent_name,
           material.c_repr(name),
           transform.c_repr(name),
           offset)

        return buf


class Cube(Shape):
    def __init__(self, yaml_obj):
        Shape.__init__(self, yaml_obj)

    @classmethod
    def from_yaml(cls, obj) -> 'Cube':
        return cls(yaml_obj=obj)

    def c_repr(self, name, parent_name, offset, resources):
        material = Material.from_yaml(self.yaml_obj['material'])
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        buf = """
    {2}
    {3}
    Shape shape_{0} = {1} + {4};
    cube(shape_{0});
    shape_set_material(shape_{0}, material_{0});
    shape_set_transform(shape_{0}, transform_{0});
""".format(name,
           parent_name,
           material.c_repr(name),
           transform.c_repr(name),
           offset)

        return buf


class Cone(Shape):
    def __init__(self, yaml_obj):
        Shape.__init__(self, yaml_obj)

    @classmethod
    def from_yaml(cls, obj) -> 'Cone':
        return cls(yaml_obj=obj)

    def c_repr(self, name, parent_name, offset, resources):
        material = Material.from_yaml(self.yaml_obj['material'])
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        buf = """
    {2}
    {3}
    Shape shape_{0} = {1} + {4};
    cone(shape_{0});
    shape_set_material(shape_{0}, material_{0});
    shape_set_transform(shape_{0}, transform_{0});
""".format(name,
           parent_name,
           material.c_repr(name),
           transform.c_repr(name),
           offset)

        if 'min' in self.yaml_obj:
            buf += "    shape_{0}->fields.cone.minimum = {1:.10f};\n".format(name, self.yaml_obj['min'])
        if 'max' in self.yaml_obj:
            buf += "    shape_{0}->fields.cone.maximum = {1:.10f};\n".format(name, self.yaml_obj['max'])
        if 'closed' in self.yaml_obj:
            bool_str = 'false'
            if self.yaml_obj['closed']:
                bool_str = 'true'
            buf += "    shape_{0}->fields.cone.closed = {1};\n".format(name, bool_str)

        return buf


class Cylinder(Shape):
    def __init__(self, yaml_obj):
        Shape.__init__(self, yaml_obj)

    @classmethod
    def from_yaml(cls, obj) -> 'Cylinder':
        b = True
        if 'closed' in obj:
            b = obj['closed']
        return cls(yaml_obj=obj)

    def c_repr(self, name, parent_name, offset, resources):
        material = Material.from_yaml(self.yaml_obj['material'])
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        buf = """
    {2}
    {3}
    Shape shape_{0} = {1} + {4};
    cylinder(shape_{0});
    shape_set_material(shape_{0}, material_{0});
    shape_set_transform(shape_{0}, transform_{0});
""".format(name,
           parent_name,
           material.c_repr(name),
           transform.c_repr(name),
           offset)

        if 'min' in self.yaml_obj:
            buf += "    shape_{0}->fields.cylinder.minimum = {1:.10f};\n".format(name, self.yaml_obj['min'])
        if 'max' in self.yaml_obj:
            buf += "    shape_{0}->fields.cylinder.maximum = {1:.10f};\n".format(name, self.yaml_obj['max'])
        if 'closed' in self.yaml_obj:
            bool_str = 'false'
            if self.yaml_obj['closed']:
                bool_str = 'true'
            buf += "    shape_{0}->fields.cylinder.closed = {1};\n".format(name, bool_str)

        return buf


class Triangle(Shape):
    def __init__(self, yaml_obj):
        Shape.__init__(self, yaml_obj)

    @classmethod
    def from_yaml(cls, obj) -> 'Triangle':
        return cls(yaml_obj=obj)

    def c_repr(self, name, parent_name, offset, resources):
        if 'material' not in self.yaml_obj:
            self.yaml_obj['material'] = {}
        if 'transform' not in self.yaml_obj:
            self.yaml_obj['transform'] = []
        material = Material.from_yaml(self.yaml_obj['material'])
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        buf = """
    {1}
    {2}
    Point shape_{0}_p1 = point({3:.10f}, {4:.10f}, {5:.10f});
    Point shape_{0}_p2 = point({6:.10f}, {7:.10f}, {8:.10f});
    Point shape_{0}_p3 = point({9:.10f}, {10:.10f}, {11:.10f});

    Shape shape_{0} = {12} + {13};
    triangle(shape_{0}, shape_{0}_p1->arr, shape_{0}_p2->arr, shape_{0}_p3->arr);
    shape_set_material(shape_{0}, material_{0});
    shape_set_transform(shape_{0}, transform_{0});

    point_free(shape_{0}_p3);
    point_free(shape_{0}_p2);
    point_free(shape_{0}_p1);
""".format(name,
          material.c_repr(name),
          transform.c_repr(name),
          self.yaml_obj['p1'][0], self.yaml_obj['p1'][1], self.yaml_obj['p1'][2],
          self.yaml_obj['p2'][0], self.yaml_obj['p2'][1], self.yaml_obj['p2'][2],
          self.yaml_obj['p3'][0], self.yaml_obj['p3'][1], self.yaml_obj['p3'][2],
          parent_name, offset)

        return buf

class SmoothTriangle(Shape):
    def __init__(self, yaml_obj):
        Shape.__init__(self, yaml_obj)

    @classmethod
    def from_yaml(cls, obj) -> 'SmoothTriangle':
        return cls(yaml_obj=obj)

    def c_repr(self, name, parent_name, offset, resources):
        if 'material' not in self.yaml_obj:
            self.yaml_obj['material'] = {}
        if 'transform' not in self.yaml_obj:
            self.yaml_obj['transform'] = []
        material = Material.from_yaml(self.yaml_obj['material'])
        transform = Transform.from_yaml(self.yaml_obj['transform'])
        buf = """
    {1}
    {2}
    Point shape_{0}_p1 = point({3:.10f}, {4:.10f}, {5:.10f});
    Point shape_{0}_p2 = point({6:.10f}, {7:.10f}, {8:.10f});
    Point shape_{0}_p3 = point({9:.10f}, {10:.10f}, {11:.10f});
    Vector shape_{0}_n1 = vector({12:.10f}, {13:.10f}, {14:.10f});
    Vector shape_{0}_n2 = vector({15:.10f}, {16:.10f}, {17:.10f});
    Vector shape_{0}_n3 = vector({18:.10f}, {19:.10f}, {20:.10f});

    Shape shape_{0} = {21} + {22};
    smooth_triangle(shape_{0}, shape_{0}_p1->arr, shape_{0}_p2->arr, shape_{0}_p3->arr, shape_{0}_n1->arr, shape_{0}_n2->arr, shape_{0}_n3->arr);
    shape_set_material(shape_{0}, material_{0});
    shape_set_transform(shape_{0}, transform_{0});

    vector_free(shape_{0}_n3);
    vector_free(shape_{0}_n2);
    vector_free(shape_{0}_n1);
    point_free(shape_{0}_p3);
    point_free(shape_{0}_p2);
    point_free(shape_{0}_p1);
    
""".format(name,
          material.c_repr(name),
          transform.c_repr(name),
          self.yaml_obj['p1'][0], self.yaml_obj['p1'][1], self.yaml_obj['p1'][2],
          self.yaml_obj['p2'][0], self.yaml_obj['p2'][1], self.yaml_obj['p2'][2],
          self.yaml_obj['p3'][0], self.yaml_obj['p3'][1], self.yaml_obj['p3'][2],
          self.yaml_obj['n1'][0], self.yaml_obj['n1'][1], self.yaml_obj['n1'][2],
          self.yaml_obj['n2'][0], self.yaml_obj['n2'][1], self.yaml_obj['n2'][2],
          self.yaml_obj['n3'][0], self.yaml_obj['n3'][1], self.yaml_obj['n3'][2],
          parent_name, offset)

        return buf


def allocate_shapes(list_of_shapes):
    num = len(list_of_shapes)
    resources = {}
    buf = """    /* shapes */
    Shape all_shapes = array_of_shapes({0});
""".format(num)
    for i, shape in enumerate(list_of_shapes):
        buf += """
    /* shape {0} */
{1}
    /* end shape {0} */""".format(i, shape.c_repr(i, "all_shapes", i, resources))
    file_name = shape.get_file_name()
    if file_name is not None:
        resources[file_name] = 'shape_{0}'.format(i)

    buf += """
    /* end shapes */
"""

    return buf

File: src/shapes/shapes.c
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "../libs/linalg/linalg.h"
#include "../pattern/pattern.h"
#include "../material/material.h"
#include "../renderer/ray.h"

#include "group.h"
#include "shapes.h"

Shape
array_of_shapes(size_t num)
{
    return array_of_shapes_realloc(NULL, num);
}

Shape
array_of_shapes_realloc(Shape ptr, size_t num)
{
    return (Shape)realloc(ptr, num * sizeof(struct shape));
}

void
shape_free(Shape s)
{
    if (s != NULL) {
        intersections_free(s->xs);
        s->xs = NULL;
        material_free(s->material);
        s->material = NULL;
        s->bbox_valid = false;
        if (s->type == SHAPE_GROUP) {
            group_free(s);
        }
        // TODO CSG free
    }
}

Intersections
shape_intersect(Shape sh, Ray r, bool stop_after_first_hit)
{
    struct ray transformed_ray;
    Ray r2 = &transformed_ray;

    if (sh->transform_identity) {
        r2 = r;
    } else {
        ray_transform(r, sh->transform_inverse, &transformed_ray);
    }

    Intersections xs = sh->local_intersect(sh, r2, stop_after_first_hit);

    return xs;
}

/*
 * https://docs.unity3d.com/Manual/StandardShaderMaterialParameterNormalMap.html
 * explains how to apply a normal map stored as an RGB image.
 */
void
shape_normal_at(Shape sh, Point world_point, Intersection hit, Vector res)
{
    Point local_point;
    Vector local_normal;
    Vector perturbed;
    Vector world_normal;
    Vector tmp;

    sh->world_to_object(sh, world_point, local_point);
    sh->local_normal_at(sh, local_point, hit, local_normal);
    sh->normal_to_world(sh, local_normal, world_normal);

    if (sh->material->map_bump != NULL) {
        sh->material->map_bump->pattern_at_shape(sh->material->map_bump, sh, world_point, tmp);

        vector_scale(tmp, 2.0);
        perturbed[0] = tmp[0] - 1.0;
        perturbed[1] = tmp[1] - 1.0;
        perturbed[2] = tmp[2] - 1.0;
        perturbed[3] = 0.0;

        //vector_scale(world_normal, 0.5);
        color_accumulate(world_normal, perturbed);
    }

    vector_normalize(world_normal, res);
}

void
shape_normal_to_world(Shape sh, Vector local_normal, Vector res)
{
    Matrix tr;
    Vector un_normal;
    Vector normal;

    matrix_transpose(sh->transform_inverse, tr);

    if (sh->transform_identity) {
        vector_copy(normal, local_normal);
    } else {
        matrix_vector_multiply(tr, local_normal, un_normal);
        vector_normalize(un_normal, normal);
    }

    Vector tmp;
    if (sh->parent != NULL) {
        sh->parent->normal_to_world(sh->parent, normal, tmp);
    } else {
        vector_copy(tmp, normal);
    }
    vector_copy(res, tmp);
}

void
shape_world_to_object(Shape sh, Point pt, Point res)
{
    Point  tmp;
    if (sh->parent != NULL) {
        sh->parent->world_to_object(sh->parent, pt, tmp);
    } else {
        point_copy(tmp, pt);
    }

    if (sh->transform_identity) {
        point_copy(res, pt);
    } else {
        matrix_point_multiply(sh->transform_inverse, tmp, res);
    }
}

void
shape_divide(Shape sh, size_t threshold)
{
    return;
}

bool
shape_includes(Shape a, Shape b)
{
    return a == b;
}

void
shape_set_transform(Shape obj, const Matrix m)
{
    if (obj != NULL) {
        matrix_copy(m, obj->transform);
        matrix_inverse(m, obj->transform_inverse);
        obj->transform_identity = matrix_equal(m, MATRIX_IDENTITY);
    }
}

void
shape_set_material(Shape obj, Material m)
{
    if (obj != NULL) {
        if (obj->material == m) {
            return;
        }
        if (obj->material != NULL) {
            material_free(obj->material);
        }
        obj->material = m;
        if (m != NULL) {
            m->ref_count++;
        }
    }
}

void
shape_set_material_recursive(Shape obj, Material m)
{
    if (obj != NULL) {
        shape_set_material(obj, m);
        if (obj->type == SHAPE_GROUP) {
            int i;
            for (i = 0; i< obj->fields.group.num_children; ++i) {
                shape_set_material_recursive(obj->fields.group.children + i, m);
            }
        }
    }
}

// bounding boxes
void
shape_bounds(Shape sh, Bounding_box *res)
{
    if (!sh->bbox_valid) {
        sh->bbox_valid = true;
        double arr[4] = {-1.0, -1.0, -1.0, 1.0};
        bounding_box_add_array(&(sh->bbox), arr);

        arr[0] = 1.0;
        arr[1] = 1.0;
        arr[2] = 1.0;
        bounding_box_add_array(&(sh->bbox), arr);

        bounding_box_transform(&(sh->bbox), sh->transform, &(sh->bbox_inverse));
    }

    *res = sh->bbox;
}

void
shape_parent_space_bounds(Shape sh, Bounding_box *res)
{
    if (!sh->bbox_valid) {
        sh->bounds(sh, res); // instead of declaring a stack variable
    }

    *res = sh->bbox_inverse;
}

int
shape_to_string(char *buf, size_t n, Shape sh)
{
    const char *type_name;
    switch (sh->type) {
    case SHAPE_CONE:
        type_name = "cone";
        break;
    case SHAPE_CUBE:
        type_name = "cube";
        break;
    case SHAPE_CYLINDER:
        type_name = "cylinder";
        break;
    case SHAPE_PLANE:
        type_name = "plane";
        break;
    case SHAPE_SMOOTH_TRIANGLE:
        type_name = "smooth triangle";
        break;
    case SHAPE_SPHERE:
        type_name = "sphere";
        break;
    case SHAPE_TOROID:
        type_name = "toroid";
        break;
    case SHAPE_TRIANGLE:
        type_name = "triangle";
        break;
    case SHAPE_CSG:
        type_name = "csg";
        break;
    case SHAPE_GROUP:
        type_name = "group";
        break;
    }

    return snprintf(buf, n, "Shape:\n\ttransform: %p\n\tmaterial: %p\n\tparent: %p\n\ttype: %s\n",
        (void *)sh->transform,
        (void *)sh->material,
        (void *)sh->parent,
        type_name);
}


void
shape_recursive_parent_update(Shape sh, Shape parent)
{
    Shape child;
    int i;

    sh->parent = parent;

    if (sh->type == SHAPE_GROUP) {
        for (i = 0, child = sh->fields.group.children;
                i < sh->fields.group.num_children;
                i++, child++) {
            shape_recursive_parent_update(child, sh);
        }
    } else if (sh->type == SHAPE_CSG) {
        shape_recursive_parent_update(sh->fields.csg.left, sh);
        shape_recursive_parent_update(sh->fields.csg.right, sh);
    }
}

void
invalidate_bounding_box(Shape sh)
{
    sh->bbox_valid = false;
    bounding_box(&(sh->bbox));
}

void
shape_recursive_invalidate_bounding_box(Shape sh)
{
    Shape child;
    int i;

    invalidate_bounding_box(sh);

    if (sh->type == SHAPE_GROUP) {
        for (i = 0, child = sh->fields.group.children;
                i < sh->fields.group.num_children;
                i++, child++) {
            shape_recursive_invalidate_bounding_box(child);
        }
    } else if (sh->type == SHAPE_CSG) {
        shape_recursive_invalidate_bounding_box(sh->fields.csg.left);
        shape_recursive_invalidate_bounding_box(sh->fields.csg.right);
    }
}
